golem-component-service-base/src/repo/component.rs

```diff
@@ -0,0 +1,648 @@
+}
+
+impl ComponentRecord {
+    pub fn new<Namespace: Display>(
+        namespace: Namespace,
+        component: Component,
+    ) -> Result<Self, String> {
+        let metadata = record_metadata_serde::serialize(&component.metadata)?;
+        Ok(Self {
+            namespace: namespace.to_string(),
+            component_id: component.versioned_component_id.component_id.0,
+            name: component.component_name.0,
+            size: component.component_size as i32,
+            version: component.versioned_component_id.version as i64,
+            user_component: component.versioned_component_id.slug(),
+            protected_component: component.protected_component_id.slug(),
+            protector_version: None,
+            metadata: metadata.into(),
+        })
+    }
+}
+
+#[async_trait]
+pub trait ComponentRepo {
+    async fn create(&self, component: &ComponentRecord) -> Result<(), RepoError>;
+
+    async fn get(
+        &self,
+        namespace: &str,
+        component_id: &Uuid,
+    ) -> Result<Vec<ComponentRecord>, RepoError>;
+
+    async fn get_all(&self, namespace: &str) -> Result<Vec<ComponentRecord>, RepoError>;
+
+    async fn get_latest_version(
+        &self,
+        namespace: &str,
+        component_id: &Uuid,
+    ) -> Result<Option<ComponentRecord>, RepoError>;
+
+    async fn get_by_version(
+        &self,
+        namespace: &str,
+        component_id: &Uuid,
+        version: u64,
+    ) -> Result<Option<ComponentRecord>, RepoError>;
+
+    async fn get_by_name(
+        &self,
+        namespace: &str,
+        name: &str,
+    ) -> Result<Vec<ComponentRecord>, RepoError>;
+
+    async fn get_id_by_name(&self, namespace: &str, name: &str) -> Result<Option<Uuid>, RepoError>;
+
+    async fn get_namespace(&self, component_id: &Uuid) -> Result<Option<String>, RepoError>;
+
+    async fn delete(&self, namespace: &str, component_id: &Uuid) -> Result<(), RepoError>;
+}
+
+pub struct DbComponentRepo<DB: Database> {
+    db_pool: Arc<Pool<DB>>,
+}
+
+impl<DB: Database> DbComponentRepo<DB> {
+    pub fn new(db_pool: Arc<Pool<DB>>) -> Self {
+        Self { db_pool }
+    }
+}
+
+#[async_trait]
+impl ComponentRepo for DbComponentRepo<sqlx::Sqlite> {
+    async fn create(&self, component: &ComponentRecord) -> Result<(), RepoError> {
+        let mut transaction = self.db_pool.begin().await?;
+
+        let result = sqlx::query("SELECT namespace FROM components WHERE component_id = $1")
+            .bind(component.component_id)
+            .fetch_optional(&mut *transaction)
+            .await?;
+
+        if let Some(result) = result {
+            let namespace: String = result.get("namespace");
+            if namespace != component.namespace {
+                return Err(RepoError::Internal("Component id not unique".to_string()));
+            }
+        } else {
+            sqlx::query(
+                r#"
+                  INSERT INTO components
+                    (namespace, component_id, name)
```

Comment:
if namespace and component_id are part of key, then you can avoid extra checks probably

Reply:
@afsalthaj there is already a unique index on `(namespace, name). And `component_id` is a PK.

We are creating a new version here (in `component_version`). And for each new version we are creating `component` record first. It might exists already (if there are other versions) or not (if we are creating the first versions) - both options are not an error.

The only problem here is if we are trying to create a new version for some component id with different namespace or name comparing to the previous version.